File: Homework/quoc_khanh/lesson_7/695.MaxAreaOfIsland.py
from turtle import width
from typing import List

WATER = 0
LAND = 1
VISITED = 2
turns = [
    [-1,  0],  # UP
    [ 1,  0],  # DOWN
    [ 0, -1],  # LEFT
    [ 0,  1],  # RIGHT
]

def isCellVisted(grid: List[List[int]], x: int, y: int) -> bool:
    return grid[x][y] == VISITED

def isLand(grid: List[List[int]], x: int, y: int) -> bool:
    return grid[x][y] == LAND

def markVisited(grid: List[List[int]], row: int, col: int):
        grid[row][col] = VISITED

class Solution:
    def __init__(self) -> None:
        self.rows = 0
        self.cols = 0

    def isCellValid(self, rowIdx: int, colIdx: int) -> bool:
        return  rowIdx >=0 and \
                colIdx >= 0 and \
                rowIdx < self.rows and \
                colIdx < self.cols


    def discover(self, grid: List[List[int]], row: int, col: int):
        area = 0
        if not isCellVisted(grid, row, col) and isLand(grid, row, col):
            markVisited(grid, row, col)
            area += 1
            for x, y in turns:
                nextRow = row + x
                nextCol = col + y
                if self.isCellValid(nextRow, nextCol):
                    area += self.discover(grid, nextRow, nextCol)
        else:
            markVisited(grid, row, col)

        return area

    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        self.rows = rows = len(grid)
        self.cols = cols = len(grid[0])
        maxArea = 0
        for row in range(rows):
            for col in range(cols):
                area = self.discover(grid, row, col)
                maxArea = max(area, maxArea)
        return maxArea

s = Solution()
grid = [
    [0,0,1,0,0,0,0,1,0,0,0,0,0],
    [0,0,0,0,0,0,0,1,1,1,0,0,0],
    [0,1,1,0,1,0,0,0,0,0,0,0,0],
    [0,1,0,0,1,1,0,0,1,0,1,0,0],
    [0,1,0,0,1,1,0,0,1,1,1,0,0],
    [0,0,0,0,0,0,0,0,0,0,1,0,0],
    [0,0,0,0,0,0,0,1,1,1,0,0,0],
    [0,0,0,0,0,0,0,1,1,0,0,0,0]
]
print(s.maxAreaOfIsland(grid) == 6)

grid = [[0,0,0,0,0,0,0,0]]
print(s.maxAreaOfIsland(grid) == 0)

File: Homework/Hoang_Nguyen/lesson7/695_maxAreaOfIsland.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
var maxAreaOfIsland = function (grid) {
  const m = grid.length;
  const n = grid[0].length;

  const visited = [];
  for (let i = 0; i < m; i++) {
    visited.push([]);
  }

  const dr = [0, 1, 0, -1];
  const dc = [1, 0, -1, 0];

  let maxArea = 0;
  let currentArea = 0;

  function visit(r, c) {
    visited[r][c] = true;

    currentArea++;
    if (currentArea > maxArea) {
      maxArea = currentArea;
    }

    for (let i = 0; i < 4; i++) {
      const r1 = r + dr[i];
      const c1 = c + dc[i];
      if (
        r1 >= 0 &&
        r1 < m &&
        c1 >= 0 &&
        c1 < n &&
        !visited[r1][c1] &&
        grid[r1][c1]
      ) {
        visit(r1, c1);
      }
    }
  }

  for (let r = 0; r < m; r++) {
    for (let c = 0; c < n; c++) {
      if (!visited[r][c] && grid[r][c]) {
        currentArea = 0;
        visit(r, c);
      }
    }
  }
  return maxArea;
};
